### Pipeline

Preprocessing, alignment, and quality control of RIP-seq/Total RNA-seq experiments were performed
using `nf-core/rna-seq` pipeline using the following commands:

```bash
cd pipeline
export NXF_VER=21.10.0
curl -s https://get.nextflow.io | bash
sudo ./nextflow -bg run nf-core/rnaseq -resume -profile docker -revision 3.4 --input=design.csv --genome=GRCm38 --aligner=hisat2
```

To start a pipeline from scratch, one needs to download raw fq.gz reads from the SRA(PRJNA767097, PRJNA767108), place
them in the `pipeline/fq.gz` folder, and modify the design document accordingly.

We also additionally filtered the resulting BAM files with the following command:

```bash
cd pipeline/results/star_salmon
mkdir final-bams
for file in *.bam
do
    samtools view -@ $(nproc) -F 2820 -b $file > final-bams/$file
done
```

### Analysis

All software dependencies are described in a Dockerfile, which can be built like this:

```bash
cd analysis
sudo docker build -t ripseq-analysis:latest docker/
# run the container
sudo docker run --rm -it -v $(pwd):/project --name ripseq-analysis-container ripseq-analysis:latest
```

Then run any **Python** script in the `scripts` folder. Additional notes for some steps are listed below.

#### DE genes

We used DESeq2 to determine genes enriched in Z22 RIP-seq over Total RNA-seq. One can run the whole analysis
using `run-DE.py` script.

Note, the analysis is based on the Salmon per-gene(transcript, in fact) quantification results. They are stored directly
in the repository and correspond to `pipeline/results/star_salmon/SAMPLE/quant.sf` files generated by the pipeline.

To augment our set of ISGs with early IFN-genes, we used RNA-seq data from `GSE118926`. The experiment was processed by
the same pipeline; Salmon counts are provided in the `analysis/resources/GSE118926` folder.

#### Alu editing index (AEI)

To compute the AEI index we used the following commands:

```bash
# Go to directory containing final-bams folder with filtered BAM files
cd pipeline/results/star_salmon/final-bams
# Reheader each BAM file (1->chr1)
for file in *.bam
do
    newfile=${file/.bam/_chr.bam}
    samtools view -H $file | sed -e 's/SN:\([0-9XY]\)/SN:chr\1/' -e 's/SN:MT/SN:chrM/' | samtools reheader - $file > $newfile
    rm -rf $file
    mv $newfile $file
done
cd ...
# build docker image
sudo docker build -t rnaedits https://raw.githubusercontent.com/a2iEditing/RNAEditingIndexer/master/Dockerfile
# run the container
sudo docker run -v $(pwd):/data --rm -it ranedits bash
# launch the pipeline
RNAEditingIndex -d /data/final-bams -l /data/logs -o /data/output -os /data/editing-summary -f _chr.bam --genome mm10 --ts $(nproc)
```

The output file located in the `summary` folder was hand-formatted and then attached as a supplement table to the GEO
submission (GSE184964).

#### Editing index for other repeats

We also calculated the editing index for other repeat families. To do this, we combined biological replicates:

```bash
cd pipeline/results/star_salmon/final-bams
samtools merge -@ $(nproc) .....
# .... merge all other experiments .....
mkdir -p ../../../../analysis/results/editing-index/bam
mv RIP-ADAR1* ../../../../analysis/results/editing-index/bam
```

Generated BED tracks for each repeat family: `python3 scripts/make-editing-index-intervals.py`.   
And started the RNAEditingIndex pipeline:

```bash
cd analysis/results/editing-index
mkdir values

sudo docker run -v $(pwd):/data --rm -it ranedits bash

for repeat in LINE LTR Low_complexity RC RNA Satellite Simple_repeat scRNA snRNA srpRNA SINE
do
    RNAEditingIndex --genome mm10 --ts $(nproc) --per_region_output --per_sample_output \ 
                    -rb /data/intervals/$repeat.bed -d /data/bam -l /data/logs \ 
                    -o /data/output -os /data/summary-$repeat -f .bam 
    rm -rf logs output
    mv /data/summary-$repeat values/$repeat
done
```

#### Mapping fragments to exons/introns/intergenic regions

To calculate a summary table showing the percentage of fragments falling into exons / introns / intergenic regions, it
is necessary to sort and place the BAM files in the `resources / BAM` folder:

```bash
cd pipeline/results/star_salmon/final-bams
for file in *.bam
do
    newfile=../../../../analysis/resources/BAM/${file/markdup.sorted.bam/.namesorted.bam}
    samtools sort -@ $(nproc) -n -o $newfile $file
done
```

Then one can run the `map-to-exons-introns-intergenic.py` script.

#### Coverage tracks

Simple CPM-normalized coverage tracks for selected genes can be built using deeptools as shown below:

```bash
# ..... -> reheader bam files as shown in the AEI section
cd pipeline/results/star_salmon/
# Example for IFNb-48h Total RNA-seq samples
samtools merge -@ $(nproc) *IFNb-48h_input*.bam IFNb-48h_input.merged.bam
samtools index -@ $(nproc) IFNb-48h_input.merged.bam

bamCoverage bamCoverage --normalizeUsing CPM -bs 5 -p 10 -b IFNb-48h_input.merged.bam -o IFNb-48h_input.merged.cpm.bw
```
